src/System.Text.JsonLab/System/Text/Json/Utf8JsonWriter.WriteProperties.cs

```diff
@@ -0,0 +1,2110 @@
+        {
+            int indent = Indentation;
+
+            // This is guaranteed not to overflow.
+            Debug.Assert(int.MaxValue - JsonWriterHelper.NewLineUtf8.Length - propertyName.Length - escapedValue.Length - 7 - indent >= 0);
+
+            // Calculated based on the following: ',\r\n  "propertyName": "value"'
+            int bytesNeeded = propertyName.Length + 7 + JsonWriterHelper.NewLineUtf8.Length + indent + escapedValue.Length;
+
+            Span<byte> byteBuffer = WritePropertyNameFormatted(propertyName, bytesNeeded, indent, out int idx);
+
+            byteBuffer[idx++] = JsonConstants.Quote;
+            escapedValue.CopyTo(byteBuffer.Slice(idx));
+            idx += escapedValue.Length;
+            byteBuffer[idx++] = JsonConstants.Quote;
+
+            _bufferWriter.Advance(idx);
+        }
+
+        public void WriteString(string propertyName, DateTime value)
+            => WriteString(propertyName.AsSpan(), value);
+
+        public void WriteString(ReadOnlySpan<char> propertyName, DateTime value)
+        {
+            JsonWriterHelper.ValidateProperty(propertyName);
+
+            WriteStringWithEncoding(MemoryMarshal.AsBytes(propertyName), value);
+        }
+
+        private void WriteStringWithEncoding(ReadOnlySpan<byte> propertyName, DateTime value)
+        {
+            if (_options.SlowPath)
+                WriteStringSlowWithEncoding(propertyName, value);
+            else
+                WriteStringFastWithEncoding(propertyName, value);
+
+            _currentDepth |= 1 << 31;
+            _tokenType = JsonTokenType.Number;
+        }
+
+        private void WriteStringFastWithEncoding(ReadOnlySpan<byte> propertyName, DateTime value)
+        {
+            // This is guaranteed not to overflow.
+            Debug.Assert(int.MaxValue - propertyName.Length / 2 * 3 - 6 - JsonConstants.MaximumDateTimeLength >= 0);
+
+            // Calculated based on the following: ',"encoded propertyName":"DateTime"'
+            int bytesNeeded = propertyName.Length / 2 * 3 + 6 + JsonConstants.MaximumDateTimeLength;
+
+            Span<byte> byteBuffer = WritePropertyNameEncoded(propertyName, bytesNeeded, out int idx);
+
+            byteBuffer[idx++] = JsonConstants.Quote;
+
+            bool result = Utf8Formatter.TryFormat(value, byteBuffer.Slice(idx), out int bytesWritten);
+            Debug.Assert(result);
+            idx += bytesWritten;
+
+            byteBuffer[idx++] = JsonConstants.Quote;
+
+            _bufferWriter.Advance(idx);
+        }
+
+        private void WriteStringSlowWithEncoding(ReadOnlySpan<byte> propertyName, DateTime value)
+        {
+            Debug.Assert(_options.Formatted || !_options.SkipValidation);
+
+            if (_options.Formatted)
+            {
+                if (!_options.SkipValidation)
+                {
+                    ValidateWritingPropertyWithEncoding(propertyName);
+                }
+                WriteStringFormattedWithEncoding(propertyName, value)
```

Comment:
Any suggestions on reducing code duplication here (or in general)? The code for `Guid`, `DateTime`, and `DateTimeOffset` is essentially identical (all calling into the `Utf8Formatter`).